More V&V, unit test, reconcile peer review

erigler-usgs opened the pull request:
I requested a technical peer review from Abram Claycomb, now at the National
Oceanic and Atmospheric Administration (NOAA). He looked through the code,
"kicked the tires", and could find no obvious technical problems with the
software. His emailed response was in fact:

>   Josh,
>
>   Actually, I was trying to come up with some constructive criticisms or
>   stuff I didn't understand.  I thought I would break it or I'd have
>   difficulty on the noaa network, but no problems.  So, I want to say
>   well done, it installs, connects, downloads, works and outputs seemingly
>   flawlessly.  If there are specific questions you had answers to that you
>   want asked, I can do that, but otherwise I'll open an 'issue' and say
>   well done, no technical questions from Abe.
>
>   Cheers,
>
>   Abe

Since Abe asked for so much, I went ahead and expanded the V&V results
presented in the SECS_Verify_Demo.ipynb notebook, tweaked the README,
regenerated output, rearranged the package directory hiearchy a bit,
and created some actual Nose-friendly unit tests.

The package is now ready for a formal administrative code review. If
you are so fortunate as to be the reviewer, please add comments to the
pull request that will get submited with this commit. -EJR

erigler-usgs committed:
More V&V, unit test, reconcile peer review

I reqquest a technical peer review from Abram Claycomb, now at the National
Oceanic and Atmospheric Administration (NOAA). He looked through the code,
"kicked the tires", and could find no obvious technical problems with the
software. His emailed response was in fact:

>   Josh,
>
>   Actually, I was trying to come up with some constructive criticisms or
>   stuff I didn't understand.  I thought I would break it or I'd have
>   difficulty on the noaa network, but no problems.  So, I want to say
>   well done, it installs, connects, downloads, works and outputs seemingly
>   flawlessly.  If there are specific questions you had answers to that you
>   want asked, I can do that, but otherwise I'll open an 'issue' and say
>   well done, no technical questions from Abe.
>
>   Cheers,
>
>   Abe

Since Abe asked for so much, I went ahead and expanded the V&V results
presented in the SECS_Verify_Demo.ipynb notebook, tweaked the README,
regenerated output, rearranged the package directory hiearchy a bit,
and created some actual Nose-friendly unit tests.

The package is now ready for a formal administrative code review. If
you are so fortunate as to be the reviewer, please add comments to the
pull request that will get submited with this commit. -EJR

erigler-usgs committed:
added docs/images subfolder

Commit:
Speed up make_impmaps.py

Only draw the basemap once, and update map elements with each new time step
(thanks G. Lucas!)

Commit:
Fix updates to s_avail/s_miss circles

Something changed in Matplotlib since I first employed the set_offsets()
method to change map elements without re-initializing the whole map. This
fixes it. Also, changed font size for legend so numbers fit.

Commit:
Update demonstration data directory

Re-ran the run_me.sh script in the data folder to regenerate all the
output from the script demonstration.

Commit:
Create download and output folders

Explicitly create all folders configured at the beginning of the script.

Commit:
Updated SECS_Verify_Demo.ipynb notebook and output

Updated the SECS_Verify_Demo notebook, re-ran all cells, and generated
new output from functional tests for gridded synthetic data.

Commit:
Python 3 fixes

A number of things were failing in Python3. Fixed.

erigler-usgs committed:
Py3 and other minor fixes

- Fixed print statements to work in Py2 and Py3
- Fixed faulty logic in endtime/starttime processing

erigler-usgs committed:
Markdown additions/updates

- added Disclaimer.md
- modified License.md with explicit CC0, updated USGS copyright policy link,
  and added language related to run time libraries and dependencies
- updated and cleaned up README.md

erigler-usgs committed:
Remove and untrack plots from notebook

I updated the notebook to generate movies and display them inline instead
of generating hundreds of individual PNGs that needed to be tracked. This
commit simply removes all the old PNGs, and stops tracking them.

erigler-usgs committed:
Updated verification and demonstration notebook

- fixed to work in both Py2 and Py3
- eliminated production of PNGs in favor of inline plots only, and
  for time series of movie maps, MP4 movies.
- general code cleaning

erigler-usgs committed:
Deleted and stopped tracking output in data folder

The data folder is intended as a work space when running the run_me.sh
script. There is no reason to track the output in this folder, which may,
and probably will, change any time the run_me.sh script is executed.

erigler-usgs committed:
Track movies generated by notebook

erigler-usgs committed:
Tweak demonstration scripts

- set dates to None in make_imp_*.py scripts so they can be used out-of-the-box
  for a real time processing pipeline
- replaced NaNs in observations with zeros; this, only when combined with an
  infinite sigma for the corresponding observation, is equivalent to leaving
  that observation out of the solution
- updated run_me.sh wrapper to only generate plots from the most recent file
  generated by the make_imp_*.py scripts.

erigler-usgs committed:
Minor bug fix in imp_io.py

- a typo introduced in the last update to imp_io.py that made it work
  with Py3 was fixed.

erigler-usgs committed:
Updated installation instructions

- instructions now include the conda-forge channel, simplifying many things;
- instructions now recommend creating a Conda virtual environment;
- instructions now broken into a "required" and an "extras" section.

erigler-usgs committed:
More Py2/Py3 compatibility fixes

Apparently I failed to test the make*.py scripts in Python3 previously.
Everything now works in both, although I worry that the fixes are not
very robust.

erigler-usgs committed:
Clean up notebook

- used %%capture magic to suppress warnings, and generate cleaner output
- merged cells
- cleaned up and better organized markdown

erigler-usgs committed:
Embed videos in markdown cell

- Animations of gridded magnetic vectors were not displaying in the output
  of the cells after placing the notebook on GitHub, which kinda defeated
  the purpose of generating the animations as a demonstration. Hopefully
  this works better on GitHub.

erigler-usgs committed:
Last attempt to get video to display in notebook

Actually, it displays in downloaded notebook, but it has not been
displaying when the notebook is rendered on GitHub. If this doesn't
work, it seems clear that GitHub is causing the problem, and we'll
have to give up on this for now.

erigler-usgs committed:
Add .gitignore to repository

erigler-usgs committed:
Add code.json metadata file to project

erigler-usgs committed:
Minor tweaks to code.json

-- small changes to code.json
-- renamed License.md and Disclaimer.md to be all CAPS

drsteve commented:
@erigler-usgs : Is this still "work in progress"? Would it be safe to assume that it's largely functional in Python 3? Your dev branch might be useful to me in the short-term, but I'm assuming there's a reason this PR hasn't been merged yet?

erigler-usgs commented:
Hi @drsteve:  I and others have used this code in both Py2 and Py3 successfully over the last couple years. The more "recent" changes were all about satisfying USGS "best practices" for an official software release. Honestly, I don't remember why this stalled as a WIP almost a year ago, so I'll need to investigate and refresh my memory before merging the pull request. Nothing substantial should change when that happens.

However, there are some algorithmic changes I have in mind that will significantly increase the code's computational efficiency, allowing for denser distributions of SECs (assuming it is the SECS code you are interested in). I plan to implement these with a new major version that will probably **only** support Py3+, and may sport a different API than the existing code.

drsteve commented:
Thanks - I'll play with this branch for now. If/when you merge to master I'll switch back to that. I'm fully migrated to Python 3, so if I hit any compatibility issues I'll file a tracker here.

By the time any API changes roll around I suspect I'll be done with the current phase of work, but hopefully I'll have the need to continue using this in the future!

Edit to add: It's probably good to update the SpacePy info in the README too. Installation should be much easier across all platforms (including Windows support across 32 and 64 bit) with our recent v2 release.

erigler-usgs commented:
@mguy-usgs, I know this is ancient, and I would be happy to find someone else if you are too busy, but do you think you might be able to give this an "administrative" review? I would like to finally finalize this v1.0 so I can move on to making some substantial upgrades, and releasing a v2.0. Seriously, no hard feelings if you are too busy to take this on right now, please just let me know. Thanks!

mhearne-usgs commented:
@erigler-usgs I can take this on - I can look at it on Monday, if that's soon enough.

erigler-usgs commented:
@mhearne-usgs, that would be fantastic. No huge rush, since I'm sure you can see this has been sitting for a while now. Also, if it helps, take a look at issue #3. This was my "reconciliation" for @mguy-usgs's initial admin review. Maybe she thought it was sufficient, and just didn't get around to merging, or giving me approval to merge.

erigler-usgs committed:
Add more functionality to SECS epsilon parameter

The epsilon parameter inside the SECS class can now:
- specify the N largest singular values to retain
- specify the N smallest singular values to retain
- specify the fraction of the largest singular value's variance
  below which to keep singular values when solving a system of
  equations

in addition to its previous ability to:
- specify the fraction of the largest singular value's variance
  above which to keep singular values when solving a system of
  equations

erigler-usgs committed:
Migrate to Python3 only

Because the geomag-algorithms package (a dependency) is now Py3 only,
we will also only support Py3. These are a few minor changes that do
this. There will likely be more. After this commit however, Python 2
compatibility will be broken.

erigler-usgs committed:
Switch map generation to Cartopy from Basemap

Basemap has not been supported for several years. Cartopy replaces
most Basemap functionality, but does so using a very different API.
There might also be a few changes here related to Py2->Py3 transition.

erigler-usgs committed:
Update .gitignore

erigler-usgs committed:
Better update to .gitignore

Also slightly tweaked /data/run_me.sh, and added /data/run_me_gp.sh

erigler-usgs committed:
Addressing numerous issues from admin code review

- Added files CODE_OF_CONDUCT.md and CONTRIBUTING.md
- Updated README.md with better installation instructions
- minor tweaks to make_svsqdist.py so that it does not
  run in real-time mode by default (better for demonstration)
- removed Basemap dependency in make_imp_gp.py (uses Cartopy for
  identical functionality)
- updated code.json
- updated setup.py

erigler-usgs commented:
@mhearne-usgs ,

I wonder if you even remember reviewing this for me, but I finally got around to addressing all of your recommendations. You submitted the actual issues to the main repo, not this branch in my fork, so my mostly brief responses are to be found there, while all the actual changes are (for now) to be found in this PR. I'll remove the WIP tag from the title, and if you are satisfied with everything I did, please consider merging this, or give me the go-ahead to merge this myself. Thanks so much!

erigler-usgs committed:
Minor tweaks to README.md

- simpler conda commands
- better link for spacepy
- explicit requirement for Py3 inserted

erigler-usgs commented:
@mhearne-usgs gave me a "Teams Chat" approval and go-ahead to finally merge this PR.

mhearne-usgs commented:
The contents of the pull request indicate that Josh has addressed my issues satisfactorily.
